src/express.re

```diff
@@ -1,43 +1,149 @@
-  external sendString : t => string => unit = "send" [@@bs.send];
-  external sendJson : t => Js.Json.t => unit = "send" [@@bs.send];
-  external sendBuffer : t => Buffer.t => unit = "send" [@@bs.send];
-  external sendArray : t => array 'a => unit = "send" [@@bs.send];
-  external json : t => Js.Json.t => unit = "" [@@bs.send] [@@ocaml.deprecated "Use sendJson instead`"];
+  external sendFile : t => string => 'a => done_ = "" [@@bs.send];
+  external sendString : t => string => done_ = "send" [@@bs.send];
+  external sendJson : t => Js.Json.t => done_ = "send" [@@bs.send];
+  external sendBuffer : t => Buffer.t => done_ = "send" [@@bs.send];
+  external sendArray : t => array 'a => done_ = "send" [@@bs.send];
+  external json : t => Js.Json.t => done_ = "" 
+    [@@bs.send] [@@ocaml.deprecated "Use sendJson instead`"];
 };
 
-module Express = {
+module Next : {
+  type content;
+
+  type t = (Js.undefined content => done_) [@bs];
+
+  let router : Js.undefined content;  
+  /** value to use as [next] callback argument to skip middleware 
+      processing. */ 
+
+  let error : Error.t => Js.undefined content;
+  /** [error e] returns the argument for [next] callback to be propagate 
+      error [e] through the chain of middleware. */
+
+} = {
+
+  type content;
+
+  type t = (Js.undefined content => done_) [@bs];
+
+  external castToContent : 'a => content = "%identity";
+
+  let router = 
+    Js.Undefined.return (castToContent "router");
+
+  let error (e:Error.t) =>
+    Js.Undefined.return (castToContent e);
+
+};
+
+module Middleware = {
+  type next = Next.t;
+  type t; 
+      /* Middleware abstract type which unified the various way an
+         Express middleware can be constructed:
+         {ul
+         {- from a {b function} using the [ofFunction]}
+         {- from a Router class}
+         {- from an App class}
+         {- from the third party middleware modules}
+         }
+         For each of the class which implements the middleware interface
+         is JavaScript, one can use the "%identity" function to cast
+         it to this type [t] */
+  type f = Request.t => Response.t => next => done_;
+  
+  external fromF: f => t = "%identity"
+    [@@ocaml.doc " Create a Middleware from a function " ]; 
+  
+  type errorF = Error.t => Request.t => Response.t => next => done_;
+  
+  external fromErrorF: errorF => t = "%identity"
+    [@@ocaml.doc " Create a Middleware from an error function " ]; 
+};
+
+/* Generate the common Middleware binding function for a given
+ * type. This Functor is used for the Router and App classes. */
+module MakeBindFunctions (T: {type t;}) => {
+  external use : T.t => Middleware.t => unit = "" [@@bs.send];
+  external useOnPath : T.t => path::string => Middleware.t => unit = "use" [@@bs.send];
+  external get : T.t => path::string => Middleware.t => unit = "" [@@bs.send];
+  external getN : T.t => path::string => array Middleware.t => unit = "get" [@@bs.send];
```

Comment:
It's not clear to me what `getN` will do with all these middlewares. Will it run them in sequence? If so, why not keep just `get` and compose the middlewares? I think the name could be improved here.

(Btw, I could check the express.js docs of course, but I'm being deliberately ignorant to try to understand this from the perspective of a beginner)